---
layout: default
title: Configuration
---

<h1>Configuration:</h1>

<p>Default configurations can be set at the following configuration file: <code>config/basic.yaml</code>. Otherwise configurations can be set manually on a run-by-run basis when running FilTar from the command line by using the <code>--use-conda</code> option passed to snakemake</p>

<p>The primary options to be configured in the configuration files are as followed:</p>
<ol>
<li><code>tissues</code>: A mapping between tissue or cell line identifiers for a given species, their respective sample accession (c.f. <a href="https://www.ncbi.nlm.nih.gov/biosample/">BioSamples</a>). This mapping is mediated by a three letter species code (<i>e.g.</i> hsa) to differentiate between the same tissue type found in different species.</li> 
<li><code>samples</code>: The mapping of sample accessions to sequencing run accessions</li>
<li><code>single_end</code> and <code>paired_end</code>: A labelling of each run accession to determine if that library corresponds to single-end or paired-end sequencing</li>
<li><code>miRNAs</code>: A list of miRNAs (using canonical miRNA names found in miRBase with the three letter species prefix) to use for target prediction. If this configuration is left blank, target prediction is performed for all miRBase annotated miRNAs of that species</li>
<li><code>transcripts</code>: A list of mRNAs to use for target prediction. If this configuration is left blank, target prediction is performed for all annotated transcripts of a given species</li>
<li><code>species</code>: When specifiying the species option, a three-letter code must be used which uses the first letter of the Genus name and the first two letters of the species name. All letters must be in lowercase. A full list of available species that can be used with FilTar is found in config/species.yaml.</li>
<li><code>context</code>: Specify which RNA-Seq experiment to use e.g. liver or kidney. This should correspond to values specified for 'tissues'.If the user does not want to use RNA-Seq data at all, and to default on standard Ensembl 3UTR models, the value of 'context' should be set to 'reference'</li>
<li><code>reannotation</code>: Set to 'True' if the user wants to reannotate 3'UTRs, if not, set to 'False'.</li>
<li><code>TPM_expression_threshold</code>: set TPM expression threshold. Use a value of 0 if you do not want to filter by expression level.</li>
<li><code>conservation</code>: Set to 'True' to use multiple sequence alignments, otherwise set to 'False'. Genome-wide alignment files are many gigabytes in size, and take longer to process than non-MSA sequence data, so in some cases, this option may be preferable</li>
<li><code>prediction_algorithm</code>: Which miRNA target prediction algorithm to use - possible values are 'Targetscan7' and 'miRanda'</li>
<li><code>use_high_conf_mirs_only</code>: Set to <code>True</code> to use high confidence miRBase miRNA annotations only. Otherwise, all miRNAs for that species will be used</li>
</ol>
<p>These details can be manually configured by editing the <code>config/basic.yaml</code> file to enter configurations in serialised format.</p>

<h2>User RNA-Seq data</h2>

<p>Users can provide their own data for use with FilTar</p>

<p>To do this, users must deposit their own compressed fastq data with <code>fastq.gz</code> or <code>fq.gz</code> file extension in either the <code>data/single_end</code> or <code>data/paired_end</code> directories. When using FilTar in this way, users must also set the <code>sequence_data_source</code> configuration value to <code>User</code></p>

<p>Files names must be in the form <code>{accession}.fastq.gz</code> for single-end data and <code>{accession}_1.fastq.gz</code> and <code>{accession}_2.fastq.gz</code> containing the first and second mate pair reads, respectively for paired-end data.</p> 

<p>Users can specify any value they wish for the accession wildcard provided that their corresponds to values specified in the main configuration file.</p>

<h2>Rule parameterisation</h2>

<p>For the most part, external tools are executed using default parameters which are documented in their respective snakefiles (found within subdirectories of the modules directory). Values of parameters can be altered by directly editing the 'param' key-value pairs contained within the rules of these snakefiles.</p>

